File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
'''
비선점, 짧은 것, 빠른 것, 번호 순

시작순으로 오름차순 정렬

일단 첫 시작은 가장 빠른 것 부터 시작
종료시간(시작 + 경과)이 되면 
종료시간 이하까지 힙에 넣기, 힙에 넣을 때는 짧,빠,번

만약 힙에 넣을 것이 없으면 그 다음 작업을 힙에 넣음

힙에 작업이 없을 때 까지
'''
from heapq import heappush, heappop
def solution(jobs):
    n = len(jobs)
    jobs.sort(key=lambda x: (x[0], x[1]))
    print(jobs)
    heap = []
    i = 0
    heappush(heap, (jobs[0][1], jobs[0][0]))
    i += 1
    time = jobs[0][0]
    times = []
    while heap:
        #print(heap)
        w_time, r_time = heappop(heap)
        # 시간 작업
        time += w_time
        times.append(time - r_time)
        #print(time)
        while i < n and jobs[i][0] <= time: # 요청시간 된 얘들 작업 큐에 넣기
            heappush(heap, (jobs[i][1], jobs[i][0]))
            i += 1
            check = False
        if not heap and i < n: # 바로 이어서 작업하지 않으면 다음 요청
            heappush(heap, (jobs[i][1], jobs[i][0]))
            time = jobs[i][0]
            i += 1
    print(times)
    return sum(times) // n
    
    
    
    
    

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;
class Task implements Comparable<Task>{
    int length;
    int time;
    int id;
    public Task(int length, int time, int id){
        this.length = length;
        this.time = time;
        this.id = id;
    }
    
    @Override
    public int compareTo(Task task){
        if(this.length != task.length){ 
            // 소요시간 비교
            return this.length - task.length;    
        }
        else{
            if(this.time != task.time){
                // 요청 시각 비교
                return this.time - task.time;
            }
            else{
                return this.id - task.id;
            }
        }
    }
    @Override
    public String toString(){
        return "id:" + this.id + " 소요시간:" + this.length + " 요청시간:" + this.time;
    }
    
}
class Solution {
    public int solution(int[][] jobs) {
        /*
        job에 id를 붙여서 Tasks로 저장
        Task를 먼저 요청 시점기준으로 오름차순 정렬
        그다음 매 1초마다 요청시점의 작업 Heap에 넣기
        
        아 근데 요청 시점이 이르면 애초에 먼저 실행되니까 굳이 정렬기준에 넣을 필요도 없을 듯?
        아니지 작업 중일 때 대기큐로 들어갈 수 있으니 정렬 기준에 유지는 해야함
        
        그냥 1초씩 지나가면서 해도 될듯 1000초니까
        */
        int reqSum = 0;
        List<Task> tasks = new ArrayList<>();  
        for(int i=0; i<jobs.length; i++){
            tasks.add(new Task(jobs[i][1], jobs[i][0], i));
            reqSum += jobs[i][0];
        }
        Collections.sort(tasks, Comparator.comparing(x -> x.time));
        // for(int i=0; i<tasks.size(); i++){
        //     System.out.println(tasks.get(i));    
        // }
        PriorityQueue<Task> heap = new PriorityQueue<>();
        // job을 넣고 job이 끝나는 시점 >= tasks의 time일 때 까지 tasks에서 poll해서 넣기
        int j=0;
        int cur = tasks.get(j).time;
        while(j < tasks.size() && tasks.get(j).time <= cur){
            heap.offer(tasks.get(j++));
        }
        int result = 0;
        Task curTask;
        while(!heap.isEmpty()){
            // 우선순위큐에서 작업 하나 꺼내서 진행하기
            curTask = heap.poll();
            //System.out.println(curTask);
            cur += curTask.length;
            result += cur;
            // 현재 시점까지 요청한 작업 우선순위 큐에 채우기
            while(j < tasks.size() && tasks.get(j).time <= cur){
                heap.offer(tasks.get(j++));
            }
            // 작업이 끝났는데 큐에 작업이 없으면 다음 작업 가져오기
            if(heap.isEmpty() && j < tasks.size()){
                Task next = tasks.get(j++);
                cur = next.time;
                heap.offer(next);
            }
        }
        //System.out.println((result - reqSum)/tasks.size());
      
        return (int)((result - reqSum)/tasks.size());
    }
}